### PR TITLE
[JENKINS-26107] Labeled block

### PR DESCRIPTION
**Pull request by jglick**
Moved from https://github.com/jenkinsci/pipeline-plugin/pull/140.

[JENKINS-26107](https://issues.jenkins-ci.org/browse/JENKINS-26107)

- [X] basic impl
- [X] test
- [X] fix `BodyInvoker.displayName`: https://github.com/jenkinsci/workflow-cps-plugin/pull/3
- [x] review appearance in _Pipeline Steps_

@reviewbybees

### CODE REVIEW

**Comment by jglick** [comment]
@HRMPW notes that the term _label_ is also used to classify and select agents, for example as a parameter to the `node` step, so its use here might be confusing. Is there an alternate term which would be intuitive as a step name, and not overloaded in a Jenkins context? _segment_?

File: src/main/java/org/jenkinsci/plugins/workflow/steps/LabelStep.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2015 Jesse Glick.
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/LabelStep.java]: 2016?

**Comment by amuniz** [comment]
:bee: - regardless the final name for this step.

**Comment by emanuelez** [comment]
FWIW I think the name `label` fits fine and I don't see a lot of confusion happening with node labels.
Looking forward to this! :)

**Comment by abayer** [comment]
I'd vote for segment or section or something like that over an already meaningful term like label.

**Comment by svanoort** [comment]
@jglick Is there a reason we don't attach a LabelAction here?  

Personally I like calling this a 'Tag' or 'NamedBlock.'   

Point of note: I am realizing that there are some serious complexities in this and in mixing and matching stages and $WhateverWeCallThis.   Stages are ordered by either iota or startTime and unscoped, but $ThisThing is scoped.   I think we should also define how we expect those 'what contains what' issues to be handled.  

A couple things to consider:

* Due to the fact that this is attached to blocks (which can nest), we may nest labelled blocks, and indeed may mix named sub-blocks with atomic steps, i.e.:

```java
label('top') {
  sh 'cp myFile newFile'
  label('more') {
    sh 'bash doMore.sh'  // Nested
      label('or_less') {  // Nested under or_less, more, then top
        sh 'less newFile'
      }
  }
  sh 'echo "we are done"'  // Nested under top
}
echo 'But wait there is more!'  //I have no label?
```

* Evaluating whether a step executed as part of a stage or a labelled block is rather complex: 
``java
stage 'mine'
echo 'I am part of the stage.'
label('no, mine!') {
  echo 'Clearly, I belong to the label.  Maybe also the stage?'
}
echo 'but who owns me? stage, presumably.'
```

Let's not even consider the case where we have a stage preceding a parallel block, with labels mixed in.

We will obviously have to store the scopes of blocks in a stage as we walk the Flow graphs to evaluate where steps ran (not a big deal), and also the list of stage steps (to see if something happened before or during one). 

Consider also: do we want to describe labels/tags/names as a *list* with all applicable values, from the outside in?  No change needed in the step itself, I think, but only in how we evaluate it within UIs. This may actually be more useful and logical for users than seeing (for example) the nearest descriptive value.  In addition it isn't the end of the world if the ordering is off sometimes.   It also avoids the puzzling behavior (for users) of seeing that last echo step as part of a stage that doesn't appear to contain the preceding blocks (since they have separate labels applied). 

Maybe I'm missing something in the intended use here?

**Comment by jglick** [comment]
> Is there a reason we don't attach a LabelAction here?

Huh? We do.

> Let's not even consider the case where we have a stage preceding a parallel block, with labels mixed in.

The intent in that case is clear to my mind. The code between one `stage` and the next serves as an implicit labeled block, enclosing the explicit ones. `FlowNodeSerialWalker` already implemented this.

The tricky case is `stage` inside some (non-`parallel`) labeled block; for that case I think we can simply say the behavior is undefined and leave it at that, since `stage` was only ever intended for top-level use.

Anyway with the advent of `label` you should just stop using `stage` anyway. There is no need to support scripts using both.

```groovy
node {
  stage 'Check out'
  checkout scm
  stage 'Build'
  sh 'make'
  stash name: 'app', includes: 'app.bin'
}
node {
  stage 'Test'
  unstash 'so'
  sh './app.bin'
}
```

should be translated to something like

```groovy
label('Dev') {
  node {
    label('Check out') {
      checkout scm
    }
    label('Build') {
      sh 'make'
      stash name: 'app', includes: 'app.bin'
    }
  }
}
label('Test') {
  node {
    unstash 'so'
    sh './app.bin'
  }
}
```

You could choose to leave the `node` blocks outside of `label`, in which case they would simply have no associated label, so any messages or pauses or metadata (for example, waiting for a free executor) would just be show at top level in some UI-specific manner, as if the entire build were wrapped in an implicit

```groovy
label('<top level>') {
  // …
}
```

> This may actually be more useful and logical for users than seeing (for example) the nearest descriptive value.

No particular opinion. Different UIs may have different needs, and different amount of space available.

**Comment by jglick** [comment]
@kohsuke wants this as a mode of `stage`.

**Comment by abayer** [comment]
+1 to that. =)

**Comment by svanoort** [comment]
It certainly reduces user confusion, but I'm wincing a bit thinking of how we'll handle pipelines with two modes of operation for stage (snapshot-in-time vs. block-scoped).

**Pull request closed without merging**